[Clang] Permit noescape on non-pointer types

Xazax-hun opened the pull request:
In modern C++ we often use span, string_view or other view objects instead of raw pointers. This PR opens the door to mark those noescape. This can be useful to document the API contracts, for interop with memory safe languages like Swift or Rust, and possibly in the future to implement take this into account in codegen.

The PR also adds a feature test macro. The goal is to help avoiding warnings when the code is compiler by earlier versions of clang that does not permit this attribute on non-pointer types.

Commit:
[Clang] Permit noescape on non-pointer types

In modern C++ we often use span, string_view or other view objects
instead of raw pointers. This PR opens the door to mark those noescape.
This can be useful to document the API contracts, for interop with
memory safe languages like Swift or Rust, and possibly in the future to
implement take this into account in codegen.

The PR also adds a feature. The goal is to help avoiding warnings when
the code is compiler by earlier versions of clang that does not permit
this attribute on non-pointer types.

erichkeane commented on clang/test/SemaCXX/noescape-attr.cpp:
```diff
@@ -1,7 +1,12 @@
\ No newline at end of file
+void test2(int __attribute__((noescape)) arr, int size);
+
+#if !__has_feature(attribute_noescape_nonpointer)
```
I don't get this macro/error here.  This will never happen in clang, we support this feature.

Xazax-hun replied:
This PR introduced this feature and the goal of this test to make sure feature testing works as expected. I am happy to remove the test if you find it redundant.

erichkeane replied:
Got it, thanks for clarifying.

erichkeane reviewed (commented):
I see you've enabled it, but the tests don't actually show that this has any effect besides suppressing a diagnostic.  I want to see something that shows that the attribute is actually DOING something on non-pointer types.

Also, requires a release note.

Xazax-hun commented:
> I want to see something that shows that the attribute is actually DOING something on non-pointer types.

I added a test to make sure the attribute is actually applied to the AST. Admittedly, there is not a lot that this can be used for as of now (other than documenting the intent that a parameter should not escape). But this is something that we want to rely on extensively in our downstream fork in Swift to support safer interop between C, C++ and Swift. I also suspect that this can be potentially useful information for generating ergonomic/safe bindings for Rust or other memory safe languages as well. 

This attribute also come up in https://discourse.llvm.org/t/rfc-a-clangir-based-safe-c/83245

Let me know if this makes sense. Otherwise, we could keep this change in our downstream fork until more work is accumulated for this attribute to do something. 

> Also, requires a release note.

Done.

erichkeane reviewed (commented):
I would still like to see a CodeGen test for this to show that it modifies the LLVM-IR.  

I see the attribute is referenced once in our codegen (`CodeGenTypes::arrangeObjCMessageSendSignature`), which looks to modify the IR, so I'd like to see a test that validates this changes with these types.

Xazax-hun commented:
>  I'd like to see a test that validates this changes with these types.

Sorry, I missed the part in CodeGen. So `noescape` is used to emit the `nocapute` parameter attribute in the LLVM IR. This parameter attribute is only valid for pointers and we currently do not seem to have a way in the LLVM IR to express nocapture for a field of a struct.

I changed the PR to make sure we do not emit `nocapture` in the IR where we should not and added some tests validating this.

erichkeane commented:
> > I'd like to see a test that validates this changes with these types.
> 
> Sorry, I missed the part in CodeGen. So `noescape` is used to emit the `nocapute` parameter attribute in the LLVM IR. This parameter attribute is only valid for pointers and we currently do not seem to have a way in the LLVM IR to express nocapture for a field of a struct.
> 
> I changed the PR to make sure we do not emit `nocapture` in the IR where we should not and added some tests validating this.

Thanks! I guess this leaves me confused as to what the PURPOSE of this patch is if it doesn't allow the attribute to get to IR?  What is it that consumes this?  Do we have some sort of analysis tool that works on it?  If so, can we write a test that shows that behavior?

Xazax-hun commented:
> What is it that consumes this?

We are planning to consume this downstream in the Swift compiler for C++ interop. Swift has non-escapable types, like Swift's `Span` type. The interop layer cannot bridge a C++ `std::span` as Swift's `Span` without this additional information from the API's authors. 

The need to annotate this property was also raised in https://discourse.llvm.org/t/rfc-a-clangir-based-safe-c/83245 which was the main motivation to make this change upstream, to avoid having multiple different spellings across the parallel efforts to interoperate between safe and unsafe code.

AaronBallman commented:
> We are planning to consume this downstream in the Swift compiler for C++ interop. Swift has non-escapable types, like Swift's `Span` type. The interop layer cannot bridge a C++ `std::span` as Swift's `Span` without this additional information from the API's authors.

Why should this live upstream if it basically only benefits a single downstream?

> The need to annotate this property was also raised in https://discourse.llvm.org/t/rfc-a-clangir-based-safe-c/83245 which was the main motivation to make this change upstream, to avoid having multiple different spellings across the parallel efforts to interoperate between safe and unsafe code.

That RFC did not really get much support because CIR isn't anywhere near far enough along to start making plans for it, so it's not really a compelling reason to move forward with these changes yet.

Xazax-hun commented:
> Why should this live upstream if it basically only benefits a single downstream?

This is a fair question, my intuition was that if the need came up multiple times independently, it might be worth to have it upstream. But it is also fair to wait until there is an actual upstream user, or multiple downstream users. I am OK with closing this and keep it downstream until either of those happens. The only risk is downstream users diverging, which is something we can always deal with during the RFC process before something is upstreamed.

AaronBallman commented:
> > Why should this live upstream if it basically only benefits a single downstream?
> 
> This is a fair question, my intuition was that if the need came up multiple times independently, it might be worth to have it upstream. But it is also fair to wait until there is an actual upstream user, or multiple downstream users. I am OK with closing this and keep it downstream until either of those happens. The only risk is downstream users diverging, which is something we can always deal with during the RFC process before something is upstreamed.

My (slight) preference is to keep this in the downstream until we have a need upstream (or multiple downstreams need it). WDYT @erichkeane?

erichkeane commented:
> > > Why should this live upstream if it basically only benefits a single downstream?
> > 
> > 
> > This is a fair question, my intuition was that if the need came up multiple times independently, it might be worth to have it upstream. But it is also fair to wait until there is an actual upstream user, or multiple downstream users. I am OK with closing this and keep it downstream until either of those happens. The only risk is downstream users diverging, which is something we can always deal with during the RFC process before something is upstreamed.
> 
> My (slight) preference is to keep this in the downstream until we have a need upstream (or multiple downstreams need it). WDYT @erichkeane?

I'm incredibly on the fence here.  We DO have a clang-tidy pass that actually consumes this it looks (`bugprone-no-escape`), so there is potential value here.  Additionally, AST matcher folks might find this change valuable.

Seeing as the REST of it is in the CFE, I can squint and see value to having it upstream.

AaronBallman commented:
> > > > Why should this live upstream if it basically only benefits a single downstream?
> > > 
> > > 
> > > This is a fair question, my intuition was that if the need came up multiple times independently, it might be worth to have it upstream. But it is also fair to wait until there is an actual upstream user, or multiple downstream users. I am OK with closing this and keep it downstream until either of those happens. The only risk is downstream users diverging, which is something we can always deal with during the RFC process before something is upstreamed.
> > 
> > 
> > My (slight) preference is to keep this in the downstream until we have a need upstream (or multiple downstreams need it). WDYT @erichkeane?
> 
> I'm incredibly on the fence here. We DO have a clang-tidy pass that actually consumes this it looks (`bugprone-no-escape`), so there is potential value here. Additionally, AST matcher folks might find this change valuable.

I don't see a whole lot of benefit to AST matcher folks, but perhaps I'm missing something?

CC @PiotrZSL @HerrCai0907 @5chmidti for opinions from clang-tidy folks on whether `bugprone-no-escape` would want to see these changes or not.

efriedma-quic commented:
I'm not sure I understand the intended meaning here.  Is noescape on a struct argument supposed to recursively apply to every pointer field of the struct?  If that's the intended meaning, should we restrict this to structs that have pointer fields?

Xazax-hun commented:
> If that's the intended meaning, should we restrict this to structs that have pointer fields?

Yeah, that is the intended meaning. I am on the fence about restricting to structs with pointers though, mostly to more ergonomically support use cases like:

```
struct SometimesHasPtr {
#ifdef COMPILE_TIME_FEATURE_FLAG
int *ptr;
#endif
};
```

A diagnostic to warn on structs that have no pointers would trigger on `SometimesHasPtr` in some build configurations but not in others. If you think that is not a blocker, I can add a diagnostic.

5chmidti commented:
> whether `bugprone-no-escape` would want to see these changes or not.

The `bugprone-no-escape` check seems to be a bit limited because it only works with `dispatch_async` and `dispatch_after` (from Apple, it seems). It could probably be expanded to include some more async constructs, but adding support for other variable types looks like it would be a positive that would immediately work for the above functions. Only the diagnostic of the check would need to be adjusted a bit.

hnrklssn commented:
The consensus in https://discourse.llvm.org/t/rfc-updating-the-semantics-of-the-noescape-attribute/90326/ seems to be that the community is okay with resurrecting this initiative :)
